gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactorJobRunner.java

```diff
@@ -397,4 +418,97 @@ public Status status() {
+          if (path.getName().endsWith(validExtention)) {
+            return true;
+          }
+        }
+        return false;
+      }
+    })) {
+      paths.add(fileStatus.getPath());
+    }
+    return paths;
+  }
+
+  /**
+   * Submit an event reporting new and cumulative late event counts.
+   */
+  private void submitLateEvent(List<Path> newLateFilePaths, Path lateDataDir) {
+    try {
+      HdfsRecordCountProvider newLateRecordCountProvider =
+          (HdfsRecordCountProvider) Class
+              .forName(
+                  this.jobProps.getProp(MRCompactor.COMPACTION_HDFS_EVENT_COUNT_PROVIDER,
+                      MRCompactor.DEFAULT_COMPACTION_HDFS_EVENT_COUNT_PROVIDER)).getConstructor(State.class)
+              .newInstance(this.jobProps);
+
+      HdfsRecordCountProvider lateDataDirRecordCountProvider =
+          (HdfsRecordCountProvider) Class
+              .forName(
+                  this.jobProps.getProp(MRCompactor.COMPACTION_HDFS_EVENT_COUNT_PROVIDER,
+                      MRCompactor.DEFAULT_COMPACTION_HDFS_EVENT_COUNT_PROVIDER)).getConstructor(State.class)
+              .newInstance(this.jobProps);
+
+      if (newLateRecordCountProvider instanceof RecordCountFromFileName) {
+        FileNameWithRecordCountFormat newLateRecordFileFormat =
+            ((RecordCountFromFileName) newLateRecordCountProvider).getFileNameFormat();
+        ((RecordCountFromFileName) lateDataDirRecordCountProvider).setFileNameFormat(new FileNameFormatWithSuffix(
+            Optional.of(newLateRecordFileFormat)));
+      }
+
+      Map<String, String> eventMetadataMap = Maps.newHashMap();
+      long newLateEventCount = newLateRecordCountProvider.getRecordCount(fs, newLateFilePaths);
+      eventMetadataMap.put("newLateEventCounts", Long.toString(newLateEventCount));
+      eventMetadataMap.put(
+          "cumulativeLateEventCounts",
+          Long.toString(newLateEventCount
+              + lateDataDirRecordCountProvider.getRecordCount(fs, this.getCumulativeLateFilePaths(lateDataDir))));
+
+      LOG.info("Submitting late event counts: " + eventMetadataMap);
+      this.eventSubmitter.submit("LateEventCounts", eventMetadataMap);
+    } catch (Exception e) {
+      LOG.error("Failed to submit late event count:" + e, e);
+    }
+  }
+
+  @NoArgsConstructor
+  @AllArgsConstructor
+  public static class FileNameFormatWithSuffix implements FileNameWithRecordCountFormat {
+    private static final String SEPARATOR = ".";
+
+    private Optional<FileNameWithRecordCountFormat> formatWithoutSuffix;
+
+    @Override
+    public String constructFileName(String... components) {
+      StringBuilder sb = new StringBuilder();
+      for (int i = 0; i < components.length; i++) {
+        if (i > 0) {
+          sb.append(SEPARATOR);
+        }
+        sb.append(components[i]);
+      }
+      return sb.toString();
+    }
+
+    @Override
+    public long getRecordCount(String fileName) {
```

Comment:
I don't understand what's going on in this method. It seems to have some implicit assumptions about what the components are but those are not documented.